include/ginkgo/core/log/profiler_hook.hpp

```diff
@@ -0,0 +1,249 @@
+/*******************************<GINKGO LICENSE>******************************
+Copyright (c) 2017-2023, the Ginkgo authors
+All rights reserved.
+
+Redistribution and use in source and binary forms, with or without
+modification, are permitted provided that the following conditions
+are met:
+
+1. Redistributions of source code must retain the above copyright
+notice, this list of conditions and the following disclaimer.
+
+2. Redistributions in binary form must reproduce the above copyright
+notice, this list of conditions and the following disclaimer in the
+documentation and/or other materials provided with the distribution.
+
+3. Neither the name of the copyright holder nor the names of its
+contributors may be used to endorse or promote products derived from
+this software without specific prior written permission.
+
+THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS
+IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED
+TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A
+PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+******************************<GINKGO LICENSE>*******************************/
+
+#ifndef GKO_PUBLIC_CORE_LOG_PROFILER_HOOK_HPP_
+#define GKO_PUBLIC_CORE_LOG_PROFILER_HOOK_HPP_
+
+
+#include <unordered_map>
+
+
+#include <ginkgo/config.hpp>
+#include <ginkgo/core/log/logger.hpp>
+
+
+namespace gko {
+namespace log {
+
+
+/** Categorization of logger events. */
+enum class profile_event_category {
+    /** Memory allocation. */
+    memory,
+    /** Kernel execution and data movement. */
+    operation,
+    /** PolymorphicObject events. */
+    object,
+    /** LinOp events. */
+    linop,
+    /** LinOpFactory events. */
+    factory,
+    /** Stopping criterion events. */
+    criterion,
+    /** For development use. */
+    internal,
+};
```

Comment:
Can we maybe unify this with the logger events? This seems to be a subset of the logger events.

Reply:
Can you elaborate on what unify means here? Create a mapping between events and categories? That wouldn't improve much, since the event types are not available in the `on_*` calls

Reply:
I mean the logger masks, defined in [core/log/logger.cpp](https://github.com/ginkgo-project/ginkgo/blob/develop/core/log/logger.cpp#L40).

Reply:
Someone using a custom profiler, would also need to use these categories, right ? Given that the use cases of these masks and the categories are essentially the same and given that ProfilerHook is in the `log` namespace, wouldn't it be cleaner to use the same masks for the ProfilerHook as well ?

Reply:
The use cases for these two types differ IMO: The values for profile_event_category are contiguous without any meaning attached to their values, by offsetting them by a large constant, I make it pretty unlikely they will collide with applications' categories. The values for the masks have specific meanings, but are all over the place, which makes it harder to avoid these kinds of NVTX category ID collisions.

Reply:
I see. I didnt consider that there might be collision issues. Do they have some recommendations on which IDs are safe to use ?

Reply:
I just offset the enum values by a magic number ("GKO" in ASCII :smile: ), that will most likely work